### PR TITLE
Turn on `ReadyToRun`

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Turn on ReadyToRun for PowerShell dll-s.

We can use https://github.com/dotnet/runtime/issues/35062#issuecomment-615005834 to check dll-s.

## PR Context

Now all .Net dll-s is in ReadyToRun format and PowerShell distributive doubled in size. So we can turn on the feature too. It will add ~20Mb (10%).

Docs https://docs.microsoft.com/en-us/dotnet/core/whats-new/dotnet-core-3-0#readytorun-images

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by TravisEz13** [comment]
We should disable the crossgen, but only for our binaries... let me see if I can do this change...

**Review by TravisEz13 — APPROVED** [review]

**Comment by TravisEz13** [comment]
I need to do some testing.

**Comment by TravisEz13** [comment]
@daxian-dbw FYI

File: build.psm1 [hunk 1]
```diff
@@ -2361,29 +2361,7 @@ function Start-CrossGen {
...
-            "Microsoft.WSMan.Management.dll",
-            "Microsoft.WSMan.Runtime.dll",
-            "Microsoft.PowerShell.Commands.Diagnostics.dll",
-            "Microsoft.PowerShell.GraphicalHost.dll",
-            "Microsoft.Management.Infrastructure.CimCmdlets.dll"
-        )
-    }
-
-    $fullAssemblyList = $commonAssembliesForAddType + $psCoreAssemblyList
+    $fullAssemblyList = $commonAssembliesForAddType
```
Comment (iSazonov) [inline build.psm1:2364]: It seems we should disable Start-CrossGen function.

Comment (SteveL-MSFT) [reply build.psm1:2364]: We should probably disable `Start-CrossGen` until we can verify PublishReadyToRun is a good replacement

Comment (iSazonov) [reply build.psm1:2365]: Done.

Comment (TravisEz13) [reply build.psm1:2365]: CrossGen and ReadyToRun are the same thing...

File: build.psm1 [hunk 2]
```diff
@@ -414,10 +414,11 @@ Fix steps:
 
     # handle ResGen
     # Heuristic to run ResGen on the fresh machine
-    if ($ResGen -or -not (Test-Path "$PSScriptRoot/src/Microsoft.PowerShell.ConsoleHost/gen")) {
-        Write-Log "Run ResGen (generating C# bindings for resx files)"
-        Start-ResGen
-    }
+    # TODO: remove Start-ResGen after testing ReadyToRun
```
Comment (TravisEz13) [inline build.psm1:417]: Please revert this change.  This is the wrong PR

**Review by TravisEz13 — CHANGES_REQUESTED** [review]

File: build.psm1 (hunk 1 above)
Comment (iSazonov) [reply build.psm1:2365]: We can use `set COMPlus_ReadyToRunLogFile=log.txt` from issue I referenced above to check dll-s. Start-CrossGen does not create R2R dll-s.

File: build.psm1 (hunk 2 above)
Comment (iSazonov) [reply build.psm1:417]: Oops, sorry.

File: build.psm1 (hunk 1 above)
Comment (TravisEz13) [reply build.psm1:2368]: @iSazonov crossgen with a switch **is** what is used to generate readytorun binaries in the dotnet project

Comment (TravisEz13) [reply build.psm1:2368]: And I switched us to that method, but per DotNet we need to use the csproj if we want our symbols to work:  https://github.com/PowerShell/PowerShell/blob/master/build.psm1#L2217

Comment (TravisEz13) [reply build.psm1:2368]: But can your explain why crossgen should be disabled on the rest of the binaries?

Comment (SteveL-MSFT) [reply build.psm1:2368]: I forgot that we also crossgen assemblies not part of the build that we include in the package.  We should probably still crossgen those.

**Comment by daxian-dbw** [comment]
> Revert "Disable Start-ResGen" 
Disable Start-CrossGen

@iSazonov You reverted a commit and then applied the same commit again.

File: build.psm1 (hunk 1 above)
Comment (iSazonov) [reply build.psm1:2367]: This make sense for third-party assemblies. For PSReadline and MI we could move to R2R format.

> @iSazonov crossgen with a switch is what is used to generate readytorun binaries in the dotnet project

With `set COMPlus_ReadyToRunLogFile=log.txt` logs show that after Start-CrossGen our dll-s are not initialized as R2R but with PublishReadyToRun are initialized.

File: build.psm1 [hunk 3]
```diff
@@ -2233,6 +2233,9 @@ function Start-CrossGen {
             $CrossgenPath
         )
 
+        # TODO: remove Start-CrossGen after testing ReadyToRun.
+        return
```
Comment (TravisEz13) [inline build.psm1:2237]: This isn't true, and we should still crossgen the un-optimized binaries.

**Review by TravisEz13 — CHANGES_REQUESTED** [review]
Please address changes.  I made ALL the needed changes the the crossgen function.  Please revert back to my commit.

File: build.psm1 (hunk 3 above)
Comment (TravisEz13) [inline build.psm1:2237]:
```suggestion
```

**Review by TravisEz13 — APPROVED** [review]

**Comment by TravisEz13** [comment]
@PoshChan  Please remind me in 1 hour

**Comment by TravisEz13** [comment]
If we need to disable crossgen on the other assemblies, please submit a separate PR.  Combining items in PRs makes it difficult if we need to backport.

**Comment by PoshChan** [comment]
@TravisEz13, this is the reminder you requested 1 hour ago

**Comment by iSazonov** [comment]
> If we need to disable crossgen on the other assemblies, please submit a separate PR. Combining items in PRs makes it difficult if we need to backport.

I have only question about PSRL and MI. We could make it readytorun in their repos. If yes I can open tracking issues. Thoughts?

**Comment by TravisEz13** [comment]
I think it's bad practices in publish the readytorun in the nuget.  But I would open a new issue, and just be sure to ask that question.

**Comment by TravisEz13** [comment]
"At present, libraries can only be R2R compiled as part of an application, not for delivery as a NuGet package."
https://devblogs.microsoft.com/dotnet/announcing-net-core-3-0-preview-6/

**Comment by iSazonov** [comment]
In our publish folder we get .Net dll-s in R2R format and trace shows this. If we compiled them locally into this format, then PSRL and MI would also be converted in it. Yes? It makes me think that we get them already in R2R format and I'd expect we could get PSRL and MI in R2R format too.
